Dev

VolodymyrYurha opened the pull request:
Main task is done
Some additions:
- CLI console
- Test project

VolodymyrYurha committed:
fixed proj naming and folders structure

VolodymyrYurha committed:
basic methods

VolodymyrYurha committed:
basic CLI

VolodymyrYurha committed:
added all comands to repo and cli

VolodymyrYurha committed:
some tests

Puchkof commented on DigitalLibrary.App/CommandProcessor.cs:
```diff
@@ -0,0 +1,200 @@
+            {
+                switch (command)
+                {
+                    case "help":
```
I would suggest to move command names to constants

Puchkof commented on DigitalLibrary.Core/DigitalLibrary.Core/Repository/BookDatabase.cs:
```diff
@@ -0,0 +1,104 @@
+        public BookDatabase(IDirectoryProvider provider)
+        {
+            var directory = provider.GetDataDirectory();
+            _filePath = Path.Combine(directory, "books.json");
```
Move file name to constants or pass to constructor as parameter

Puchkof commented on DigitalLibrary.Core/DigitalLibrary.Core/Repository/BookDatabase.cs:
```diff
@@ -0,0 +1,104 @@
+            _books.Remove(book);
+        }
+
+        public void Update(int id, Book entity)
```
For such case, I would suggest to pass DTO model instead of entity for better layers decoupling

Puchkof replied:
For forward compatibility, I would suggest to pass some `UpdateBookModel model`. In case if you add new fields to Book - the client who uses your class may pass them and expect them to be updated. In case of the model - you decouple application update logic and data access layer

VolodymyrYurha committed:
resolved first 2 fixes

- the problem with UpdateModel is that i have too many responsibilites for repositry class. (DAL +  repo + and is used directly) So i either have to add one more layer (service) or leave it as it is.

Puchkof commented on DigitalLibrary.App/CommandProcessor.cs:
```diff
@@ -7,6 +7,15 @@ public class CommandProcessor
     {
         private readonly BookDatabase _db;
 
+        private const string command_help = "help";
```
For constants, I would suggest PascalCase like `HelpCommand` or whole upper case like `HELP_COMMAND`

Puchkof commented on DigitalLibrary.Core/DigitalLibrary.Core/Repository/BookDatabase.cs:
```diff
@@ -12,10 +11,10 @@ public class BookDatabase : IJsonDataBase<Book>, IBookDatabase
         private int _nextId;
 
-        public BookDatabase(IDirectoryProvider provider)
+        public BookDatabase(IDirectoryProvider provider, string filename = "books.json")
```
I would suggest not to hardcode paths like this. Its ok for pet project but not for commercial development, as this class may be used as part of public api of some NuGet package

VolodymyrYurha replied:
ok, but BookDatabase already implement IBookDatabase, so the class is still considered to provide file specifically for working with books DB. So can we consider this class as the only one for app only to work with books? I can make this class to be a singleton or filename as a property, will it be ok?

VolodymyrYurha committed:
PR fixes

VolodymyrYurha committed:
Added lock for each operator and tested

- Now app works fine with multiple requests at the same time

VolodymyrYurha committed:
PR fixes 1

VolodymyrYurha committed:
using filescoped namespaces

VolodymyrYurha committed:
Merge pull request #2 from IT-Sadok/concurrent-access

Added lock for each operator and tested
